### PR TITLE
Signature Endpoint

### LINKED ISSUE #814: Signature endpoint with real data
This endpoint is currently mocked. Will need to update it to use real data.

See the `/institutions/<institution>/filings/<period>/submissions/<submissionId>/sign` endpoint in the [API.md doc](https://github.com/cfpb/hmda-platform/blob/6ab278b3e1398a1b458049b04fd7b9273be6cd26/Documents/API.md).

### PR DESCRIPTION
**Pull request by schbetsy**
Closes #814 

Some tests depend on #861. Those specs can be un-commented once the test data issue is fixed.

### CODE REVIEW

File: persistence/src/main/scala/hmda/persistence/institutions/SubmissionPersistence.scala [hunk 1]
```diff
@@ -29,16 +29,23 @@ object SubmissionPersistence {
       event match {
         case SubmissionCreated(s) =>
           SubmissionState(s +: submissions)
+
         case SubmissionStatusUpdated(id, status) =>
-          val x = submissions.find(x => x.id == id).getOrElse(Submission())
+          val x: Submission = submissions.find(x => x.id == id).getOrElse(Submission())
```
Comment (jmarin) [inline persistence/src/main/scala/hmda/persistence/institutions/SubmissionPersistence.scala]: Is this type needed here? I think type inference takes care of figuring out this type, and I don't find this code confusing without the declaration, it's quite obvious that `x` is going to be of type `Submission`

File: persistence/src/main/scala/hmda/persistence/institutions/SubmissionPersistence.scala [hunk 2]
```diff
@@ -29,16 +29,23 @@ object SubmissionPersistence {
...
           val i = submissions.indexOf(x)
-          if (status == Signed) {
-            SubmissionState(submissions.updated(i, x.copy(status = status, end = System.currentTimeMillis())))
-          } else {
-            SubmissionState(submissions.updated(i, x.copy(status = status)))
-          }
+
+          val updatedSub: Submission = if (status == Signed) {
```
Comment (jmarin) [inline persistence/src/main/scala/hmda/persistence/institutions/SubmissionPersistence.scala]: I would rather have the logic after the `=` encapsulated by `{ }` to improve readability a bit

File: persistence/src/main/scala/hmda/persistence/processing/SubmissionFSM.scala [hunk 3]
```diff
@@ -226,10 +226,14 @@ class SubmissionFSM(submissionId: SubmissionId)(implicit val domainEventClassTag
...
       updateStatus(status)
-      goto(Signed) applying SubmissionSigned(Submission(submissionId, status))
+      goto(Signed) applying SubmissionSigned(Submission(submissionId, status)) replying Some(status)
   }
 
   when(ValidatedWithErrors) {
+    case Event(Sign, data) =>
+      val status = hmda.model.fi.Signed
+      updateStatus(status)
+      goto(Signed) applying SubmissionSigned(Submission(submissionId, status)) replying Some(status)
```
Comment (jmarin) [inline persistence/src/main/scala/hmda/persistence/processing/SubmissionFSM.scala]: It looks like these 2 `replying Some(status)` can be moved under `when(Signed)`?

File: persistence/src/main/scala/hmda/persistence/processing/SubmissionManager.scala [hunk 4]
```diff
@@ -107,9 +107,9 @@ class SubmissionManager(submissionId: SubmissionId) extends HmdaActor {
       log.info(s"Validation completed with errors for submission: ${sId.toString}")
       submissionFSM ! CompleteValidationWithErrors
 
-    case Signed(sId) =>
-      log.info(s"Submission signed: ${sId.toString}")
-      submissionFSM ! Sign
+    case Signed =>
+      log.info(s"Submission signed: ${submissionId.toString}")
+      submissionFSM forward Sign
```
Comment (jmarin) [inline persistence/src/main/scala/hmda/persistence/processing/SubmissionManager.scala]: Why the change to `forward`?

Comment (schbetsy) [inline persistence/src/main/scala/hmda/persistence/processing/SubmissionManager.scala]: For the endpoint to work as designed, it needs a response from the actor that says whether the `Sign` action was successful (so then we can know whether to give a `202` or `400` response). 

So, the API code sends the `Sign` event, the `SubmissionManager` forwards it to the FSM, then (because it was `forward`), the FSM responds back to the API code. The FSM responds `Some(Signed)` if the state transition was successful, or `None` if it wasn't.

File: persistence/src/main/scala/hmda/persistence/processing/SubmissionFSM.scala (hunk 3 above)
Comment (schbetsy) [inline persistence/src/main/scala/hmda/persistence/processing/SubmissionFSM.scala]: We use the response to know whether the state transition was successful. I don't think it can be moved.

File: persistence/src/main/scala/hmda/persistence/processing/SubmissionManager.scala (hunk 4 above)
Comment (jmarin) [inline persistence/src/main/scala/hmda/persistence/processing/SubmissionManager.scala]: Got it, sounds good

File: persistence/src/main/scala/hmda/persistence/institutions/SubmissionPersistence.scala (hunk 1 above)
Comment (schbetsy) [inline persistence/src/main/scala/hmda/persistence/institutions/SubmissionPersistence.scala]: Oops, that was probably left over from the debugging process. Fixed in a5804cd6.

File: persistence/src/main/scala/hmda/persistence/institutions/SubmissionPersistence.scala (hunk 2 above)
Comment (schbetsy) [inline persistence/src/main/scala/hmda/persistence/institutions/SubmissionPersistence.scala]: Sounds good. Updated in a5804cd6.

File: persistence/src/main/scala/hmda/persistence/processing/SubmissionFSM.scala [hunk 5]
```diff
@@ -226,10 +226,14 @@ class SubmissionFSM(submissionId: SubmissionId)(implicit val domainEventClassTag
     case Event(Sign, data) =>
       val status = hmda.model.fi.Signed
       updateStatus(status)
-      goto(Signed) applying SubmissionSigned(Submission(submissionId, status))
+      goto(Signed) applying SubmissionSigned(Submission(submissionId, status)) replying Some(status)
   }
 
   when(ValidatedWithErrors) {
+    case Event(Sign, data) =>
```
Comment (jmarin) [inline persistence/src/main/scala/hmda/persistence/processing/SubmissionFSM.scala]: It looks like `data` is not used. Use placeholder `_` for consistency

Comment (schbetsy) [inline persistence/src/main/scala/hmda/persistence/processing/SubmissionFSM.scala]: 👍  Updated in 46ba348b.

**Comment by nickgrippin** [comment]
LGTM.  Tested it thoroughly, code looks good!  I'm a little uncomfortable merging code that has the crucial tests commented out, but I understand the necessity here.
